evm-contracts/src/v0.7/dev/CompoundPriceFlaggingValidator.sol

```diff
@@ -0,0 +1,277 @@
+   * the percentage with which the difference in prices must reside within. For example:
+   * If the prices are valid within a 5% threshold, and 1 / denominator, this denominator
+   * = 20, since 1/20 = 5.
+   */
+  function setThreshold(
+    address aggregator,
+    string calldata compoundSymbol,
+    uint8 compoundDecimals,
+    uint32 compoundDeviationThresholdDenominator
+  ) 
+    public 
+    onlyOwner() 
+  {
+    CompoundFeedDetails memory compDetails = s_thresholds[aggregator];
+    compDetails.symbol = compoundSymbol;
+    compDetails.decimals = compoundDecimals;
+    compDetails.deviationThresholdDenominator = compoundDeviationThresholdDenominator;
+    s_thresholds[aggregator] = compDetails;
+    emit ThresholdUpdated(
+      aggregator,
+      compoundSymbol,
+      compoundDecimals,
+      compoundDeviationThresholdDenominator
+    );
+  }
+
+  /**
+   * @notice Check the price deviation of an array of aggregators
+   * @dev If any of the aggregators provided have an equivalent Compound Oracle feed
+   * that with a price outside of the configured deviation, this function will return them.
+   * @param aggregators address[] memory
+   * @return address[] invalid feeds
+   */
+  function check(address[] memory aggregators) public view returns (address[] memory) {
+    address[] memory invalidAggregators = new address[](aggregators.length);
+    uint256 invalidCount = 0;
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      address aggregator = aggregators[i];
+      if (isInvalid(aggregator)) {
+        invalidAggregators[invalidCount] = aggregator;
+        invalidCount++;
+      }
+    }
+
+    if (aggregators.length != invalidCount) {
+      assembly {
+        mstore(invalidAggregators, invalidCount)
+      }
+    }
+    return invalidAggregators;
+  }
+
+  /**
+   * @notice Check and raise flags for any aggregator that has an equivalent Compound
+   * Open Oracle feed with a price deviation exceeding the configured setting.
+   * @dev This contract must have write permissions on the Flags contract
+   * @param aggregators address[] memory
+   * @return address[] memory invalid aggregators
+   */
+  function update(address[] memory aggregators) public returns (address[] memory){
+    address[] memory invalidAggregators = check(aggregators);
+    s_flags.raiseFlags(invalidAggregators);
+    return invalidAggregators;
+  }
+
+  /**
+   * @notice Check the price deviation of an array of aggregators
+   * @dev If any of the aggregators provided have an equivalent Compound Oracle feed
+   * that with a price outside of the configured deviation, this function will return them.
+   * @param data bytes encoded address array
+   * @return needsUpkeep bool indicating whether upkeep needs to be performed
+   * @return invalid aggregators - bytes encoded address array of invalid aggregator addresses
+   */
+  function checkForUpkeep(bytes calldata data) external view override returns (bool, bytes memory) {
```

Comment:
```suggestion
  function checkUpkeep(bytes calldata data) external view override returns (bool, bytes memory) {
```
New name, little more consistent with `performUpkeep`.

Reply:
This will break inheritance, as we've pulled in the `UpkeepInterface` from c3.
Changing this would mean changing the interface and the Staleness validator.